Task 1/2019201419王成瑞/double.h

```diff
@@ -0,0 +1,291 @@
+
+#define BIAS                ((1 << (EXP_SIZE - 1)) - 1)
+#define DOUBLE_EXP_MAX      ((1 << EXP_SIZE) - 2 - BIAS)
+#define DOUBLE_EXP_MIN      (-BIAS)
+
+#define SPECIAL_EXP         (0x7FFu)
+#define DENORMALIZED_EXP    (0u)
+
+#define VL_1                ((uint128_t)1)
+
+#define swap(a, b)                                                      \
+        {                                                               \
+            __typeof__(a) tmp = a;                                      \
+            a = b;                                                      \
+            b = tmp;                                                    \
+        }
+
+
+typedef struct myDouble {
+    data_t data[DOUBLE_SIZE];
+} double_t;
+
+unsigned getSign(const double_t* _val) {
+    return _val -> data[DOUBLE_SIZE - 1] >> 7;
+}
+void setSign(double_t* _val, unsigned num) {
+    num &= 1u;
+    _val -> data[DOUBLE_SIZE - 1] = (data_t)((num << 7) | (_val -> data[DOUBLE_SIZE - 1] & 0x7Fu));
+}
+
+unsigned getExponent(const double_t* _val) {
+    return ((_val -> data[DOUBLE_SIZE - 1] & 0x7Fu) << 4) | (_val -> data[DOUBLE_SIZE - 2] >> 4);
+}
+void setExponent(double_t* _val, unsigned num) {
+    num &= (1u << EXP_SIZE) - 1;
+    _val -> data[DOUBLE_SIZE - 1] = (data_t)((getSign(_val) << 7) | (num >> 4));
+    _val -> data[DOUBLE_SIZE - 2] = (data_t)((num << 4) | (_val -> data[DOUBLE_SIZE - 2] & 0xFu));
+}
+
+uint64_t getFraction(const double_t* _val) {
+    uint64_t res = _val -> data[DOUBLE_SIZE - 2] & 0xFu;
+    for(int i = DOUBLE_SIZE - 3; i >= 0; i--) {
+        res = (res << 8) | _val -> data[i];
+    }
+    return res;
+}
+void setFraction(double_t* _val, uint64_t num) {
+    num &= (1ul << FRAC_SIZE) - 1;
+    for(int i = 0; i < DOUBLE_SIZE - 2; i++) {
+        _val -> data[i] = (data_t)(num & 0xFFu);
+        num >>= 8;
+    }
+    _val -> data[DOUBLE_SIZE - 2] = (data_t)((_val -> data[DOUBLE_SIZE - 2] & 0xF0u) | num);
+}
+
+int isZero(const double_t _val) {
+    if(getExponent(&_val) == 0 && getFraction(&_val) == 0) {
+        return getSign(&_val) ? 1 : -1;
+    } else {
+        return 0;
+    }
+}
+void setZero(double_t* _val) {
+    setExponent(_val, 0);
+    setFraction(_val, 0);
+}
+
+int isNan(const double_t _val) {
+    return getExponent(&_val) == SPECIAL_EXP && getFraction(&_val) != 0;
+}
+void setNan(double_t* _val) {
+    setExponent(_val, SPECIAL_EXP);
+    setFraction(_val, 1);
+}
+
+int isInf(const double_t _val) {
+    if(getExponent(&_val) == SPECIAL_EXP && getFraction(&_val) == 0) {
+        return getSign(&_val) ? 1 : -1;
+    } else {
+        return 0;
+    }
+}
+void setInf(double_t* _val) {
+    setExponent(_val, SPECIAL_EXP);
+    setFraction(_val, 0);
+}
+
+void __initialize__(double_t* _val) {
+    memset(_val, 0, sizeof(double_t));
+}
+
+int isDenormalized(const double_t _val) {
+    return getExponent(&_val) == 0;
+}
+
+#define Rounding(PRECISION)                                             \
```

Comment:
Neither a global macro with so many lines nor a internal macro defined in header is a good practice. Consider write subprograms like this as a function intead next time.